dpctl/tensor/libtensor/include/utils/sycl_alloc_utils.hpp

```diff
@@ -78,6 +79,28 @@ template <typename T> void sycl_free_noexcept(T *ptr, sycl::queue &q) noexcept
     sycl_free_noexcept(ptr, q.get_context());
 }
 
+class USMDeleter
+{
+private:
+    sycl::context ctx_;
+
+public:
+    USMDeleter(const sycl::queue &q) : ctx_(q.get_context()) {}
+    USMDeleter(const sycl::context &ctx) : ctx_(ctx) {}
+
+    template <typename T> void operator()(T *ptr)
+    {
+        sycl_free_noexcept(ptr, ctx_);
+    }
+};
+
+template <typename T>
+std::unique_ptr<T, USMDeleter> make_owner(T *ptr, const sycl::queue &q)
+{
+    auto usm_deleter = USMDeleter(q);
+    return std::unique_ptr<T, USMDeleter>(ptr, usm_deleter);
+}
```

Comment:
I implemented something similar for the `dpnp.choose` PR.

A couple of remaining issues have been mentioned there, but it might be worth taking a look.
https://github.com/IntelPython/dpnp/pull/2201

The biggest difference is that, rather than allocating and then using the `unique_ptr` as functionally a deleter, it wraps the `malloc_device` call as well. And as was pointed out in the PR, it could wrap the `std::runtime_error` if it fails to allocate, too.

We could do something similar here and factor away the `malloc_device` calls as well. Then any function calls that require a raw pointer can be passed `usm_unique_ptr_obj.get()` to make clear that a raw pointer managed by `unique_ptr` is being used.

If we did something like this, we can also expose it for dpnp and it can be removed from the PR.

Reply:
I am implemented `smart_malloc_device` to make allocation returning `unique_ptr`, and `async_smart_free` which takes the smart pointer, submits the `host_task` and releases the smart pointer.